tough-kms/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+        };
+        // Get the public key from AWS KMS
+        let fut = kms_client.get_public_key(rusoto_kms::GetPublicKeyRequest {
+            key_id: self.key_id.clone(),
+            ..rusoto_kms::GetPublicKeyRequest::default()
+        });
+        let response = tokio::runtime::Runtime::new()
+            .context(error::RuntimeCreation)?
+            .block_on(fut)
+            .context(error::KmsGetPublicKey {
+                profile: self.profile.clone(),
+                key_id: self.key_id.clone(),
+            })?;
+        let pb_key: Decoded<RsaPem> = response
+            .public_key
+            .context(error::PublicKeyNone)?
+            .to_vec()
+            .into();
+        Ok(Box::new(KmsRsaKey {
+            profile: self.profile.clone(),
+            client: Some(kms_client.clone()),
+            key_id: self.key_id.clone(),
+            public_key: pb_key,
+            signing_algorithm: KmsSigningAlgorithms::Rsa(String::from("RSASSA_PSS_SHA_256")),
+        }))
+    }
+
+    fn write(
+        &self,
+        _value: &str,
+        _key_id_hex: &str,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        Ok(())
+    }
+}
+
+/// Implements the Sign trait for KMS rsa Key
+pub struct KmsRsaKey {
+    /// Key Id of Customer Managed Key in KMS used to sign the message
+    pub key_id: String,
+    /// Aws account profile
+    pub profile: Option<String>,
+    /// KmsClient Object to query AWS KMS
+    pub client: Option<KmsClient>,
+    /// Public Key corresponding to Customer Managed Key
+    pub public_key: Decoded<RsaPem>,
+    /// Signing Algorithm to be used for the Customer Managed Key
+    pub signing_algorithm: KmsSigningAlgorithms,
+}
+
+impl fmt::Debug for KmsRsaKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("KmsRsaKey")
+            .field("key_id", &self.key_id)
+            .field("signing_algorithm", &self.signing_algorithm)
+            .field("public_key", &self.public_key)
+            .finish()
+    }
+}
+
+impl Sign for KmsRsaKey {
+    fn tuf_key(&self) -> Key {
+        // Create a Key struct for the public key
+        Key::Rsa {
+            keyval: RsaKey {
+                public: self.public_key.to_owned(),
+                _extra: HashMap::new(),
+            },
+            scheme: RsaScheme::RsassaPssSha256,
+            _extra: HashMap::new(),
+        }
+    }
+
+    fn sign(
+        &self,
+        msg: &[u8],
+        _rng: &dyn SecureRandom,
+    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let kms_client = match self.client.to_owned() {
+            Some(value) => value,
+            None => client::build_client_kms(self.profile.as_deref())?,
+        };
+        let sign_fut = kms_client.sign(SignRequest {
+            key_id: self.key_id.clone(),
+            message: bytes::Bytes::from(msg.to_owned()),
+            message_type: Some(String::from("RAW")),
```

Comment:
Apparently [messages can only be 4096 bytes long](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html#KMS-Sign-request-Message), so we will want to use digest mode here.

Along with a `use ring::digest::{digest, SHA256};` up top, I've replaced this with:

```suggestion
            message: digest(&SHA256, msg).as_ref().to_vec().into(),
            message_type: Some(String::from("DIGEST")),
```

(This doesn't fix the mysterious invalid signature issue, which I'm working through separately.)

Reply:
Done. I wanted to use "DIGEST" but couldn't figure out `digest(&SHA256, msg).as_ref().to_vec().into()` part